From 315a679854cb23b0aff994ceb471b0d44c3818e7 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Thu, 21 Jul 2022 16:05:50 +0000
Subject: [PATCH] Revert "Revert "[SW-77251] gaudi3: add valid address check api""

This api can be used to perform generic address bound checks. Hence re-reverting the patch.

This reverts commit a5f85aac5254bba1728ff3fc3c7208e80fefee0d.

Change-Id: Ie52ab1653a04f15d1354076dcb998c8f5d511017
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 80be49f..82fe316 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -894,6 +894,37 @@
 	return 0;
 }
 
+bool arcboot_is_valid_address_range(uint64_t addr, int len)
+{
+	if (addr >= SP_SRAM_BASE &&
+	    addr + len <= SP_SRAM_BASE + HL_SP_SRAM_SIZE)
+		return true;
+
+	if (addr >= HL_SP_SRAM_BASE &&
+	    addr + len <= HL_SP_SRAM_BASE + HL_SP_SRAM_SIZE)
+		return true;
+
+	if (addr >= HL_ARC0_ICCM_FULL_BASE &&
+	    addr + len <= HL_ARC0_ICCM_FULL_BASE + ARC0_MEM_SIZE)
+		return true;
+
+	if (addr >= HL_ARC1_ICCM_FULL_BASE &&
+	    addr + len <= HL_ARC1_ICCM_FULL_BASE + ARC1_MEM_SIZE)
+		return true;
+
+	if (addr >= HL_ARC2_ICCM_FULL_BASE &&
+	    addr + len <= HL_ARC2_ICCM_FULL_BASE + ARC2_MEM_SIZE)
+		return true;
+
+	if (addr >= HL_SRAM_BASE && addr + len <= HL_SRAM_BASE + HL_SRAM_SIZE)
+		return true;
+
+	if (addr >= HL_SRAM_BASE_FULL && addr + len <= HL_SRAM_BASE_FULL + HL_SRAM_SIZE)
+		return true;
+
+	return false;
+}
+
 void arcboot_rom_in_recovery(void)
 {
 	uint8_t override = 1;
diff --git a/arch/gaudi3/arc/soc.h b/arch/gaudi3/arc/soc.h
index 156bdf9..98e3061 100644
--- a/arch/gaudi3/arc/soc.h
+++ b/arch/gaudi3/arc/soc.h
@@ -38,6 +38,8 @@
 #define SPI_FLASH_BLOCK_SIZE		1024
 
 #define REGION_BASE_ADDR(REG_IDX)	(REG_IDX << 28)
+#define SZ_32K				0x00008000
+#define SZ_64K				0x00010000
 #define SZ_128K				0x00020000
 #define SZ_256K				0x00040000
 #define SZ_256MB			0x10000000
@@ -83,6 +85,7 @@
 #define HL_SCRATCHPAD_SRAM_OFFSET	0x3120000
 #define HL_FWSRAM_OFFSET		0x3140000
 #define HL_PCIE_FWSRAM_OFFSET		0x7F00000
+#define SP_SRAM_BASE			(HL_GEN_MEM_BASE + HL_SCRATCHPAD_SRAM_OFFSET)
 
 #define HL_FW_MEM_BASE_FULL		0x0300007FD0000000ull
 #define HL_GEN_MEM_BASE_FULL		HL_FW_MEM_BASE_FULL
@@ -90,9 +93,18 @@
 #define HL_SP_SRAM_SIZE			0x10000
 #define HL_MIRROR_CFG_BASE		HL_D2D_LBW_BASE
 
+#define HL_ARC0_ICCM_FULL_BASE		(HL_FW_MEM_BASE_FULL + HL_ARC0_ICCM_OFFSET)
+#define ARC0_ICCM_SIZE			SZ_128K
+#define ARC0_DCCM_SIZE			SZ_32K
 #define HL_ARC1_ICCM_FULL_BASE		(HL_FW_MEM_BASE_FULL + HL_ARC1_ICCM_OFFSET)
 #define ARC1_ICCM_SIZE			SZ_256K
 #define ARC1_DCCM_SIZE			SZ_64K
+#define HL_ARC2_ICCM_FULL_BASE		(HL_FW_MEM_BASE_FULL + HL_ARC2_ICCM_OFFSET)
+#define ARC2_ICCM_SIZE			SZ_128K
+#define ARC2_DCCM_SIZE			SZ_32K
+#define ARC0_MEM_SIZE			(ARC0_ICCM_SIZE + ARC0_DCCM_SIZE)
+#define ARC1_MEM_SIZE			(ARC1_ICCM_SIZE + ARC1_DCCM_SIZE)
+#define ARC2_MEM_SIZE			(ARC2_ICCM_SIZE + ARC2_DCCM_SIZE)
 #define HL_FW_SRAM_FULL_BASE		(HL_FW_MEM_BASE_FULL + HL_FWSRAM_OFFSET)
 #define HL_FW_SRAM_SIZE			SZ_128K
 
@@ -100,5 +112,4 @@
 #define CONV_64B_TO_ARC_SRAM(addr)	((addr & 0xFFFFFFF) + HL_SRAM_BASE)
 #define HL_CFG_SPACE_DIE1_OFFSET	BIT(23)
 
-#define CONV_64B_TO_ARC(addr)	((addr & 0xFFFFFFF) + HL_FW_MEM_BASE)
 #endif /* _SOC_H_ */
diff --git a/include/arcboot.h b/include/arcboot.h
index 927616d..1d8bb05 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -165,6 +165,7 @@
 void arcboot_rom_load_recovery(void);
 int arcboot_rom_mem_init(uint64_t addr, uint32_t size);
 int arcboot_rom_verify(enum hl_fw_boot_ptr_imgs img);
+bool arcboot_is_valid_address_range(uint64_t addr, int len);
 
 #define CLEAR_DMA_SECURITY_PROGRESS 0x3
 
